File: app/src/main/java/de/bikebean/app/db/state/LocationState.java
package de.bikebean.app.db.state;

import android.os.Bundle;

import androidx.annotation.NonNull;

import de.bikebean.app.db.DatabaseEntity;

import static de.bikebean.app.db.state.State.KEY.ACC;
import static de.bikebean.app.db.state.State.KEY.LAT;
import static de.bikebean.app.db.state.State.KEY.LNG;
import static de.bikebean.app.db.state.State.KEY.NO_CELL_TOWERS;
import static de.bikebean.app.db.state.State.KEY.NO_WIFI_ACCESS_POINTS;

public class LocationState extends DatabaseEntity {

    private final int smsId;
    private final long timestamp;

    private final double lat;
    private final double lng;
    private final double acc;

    private final int noCellTowers;
    private final int noWifiAccessPoints;

    public LocationState(final @NonNull State latState, final @NonNull State lngState,
                         final @NonNull State accState, final @NonNull State noCellTowersState,
                         final @NonNull State noWifiAccessPointsState) {
        smsId = latState.getSmsId();
        timestamp = latState.getTimestamp();

        lat = latState.getValue();
        lng = lngState.getValue();
        acc = accState.getValue();

        noCellTowers = noCellTowersState.getValue().intValue();
        noWifiAccessPoints = noWifiAccessPointsState.getValue().intValue();
    }

    public double getAcc() {
        return acc;
    }

    public double getLng() {
        return lng;
    }

    public double getLat() {
        return lat;
    }

    public long getTimestamp() {
        return timestamp;
    }

    public int getSmsId() {
        return smsId;
    }

    public @NonNull Bundle getArgs() {
        return new LocationBundle()
                .putDouble(LAT, lat)
                .putDouble(LNG, lng)
                .putDouble(ACC, acc)
                .putInt(NO_CELL_TOWERS, noCellTowers)
                .putInt(NO_WIFI_ACCESS_POINTS, noWifiAccessPoints)
                .get();
    }

    static class LocationBundle {

        private final @NonNull Bundle bundle = new Bundle();

        Bundle get() {
            return bundle;
        }

        LocationBundle putDouble(final @NonNull State.KEY key, double value) {
            bundle.putDouble(key.get(), value);
            return this;
        }

        LocationBundle putInt(final @NonNull State.KEY key, int value) {
            bundle.putInt(key.get(), value);
            return this;
        }
    }

    @Override
    public DatabaseEntity getNullType() {
        return null;
    }

    @Override
    public String createReportTitle() {
        return null;
    }

    @Override
    public String createReport() {
        return null;
    }
}

File: app/src/main/java/de/bikebean/app/db/state/LocationState.kt
package de.bikebean.app.db.state

import android.os.Bundle
import de.bikebean.app.db.DatabaseEntity
import de.bikebean.app.db.state.State.KEY
import de.bikebean.app.ui.utils.date.DateUtils
import de.bikebean.app.ui.utils.date.DateUtils.getDaysSinceState

class LocationState(
        latState: State,
        lngState: State,
        accState: State,
        noCellTowersState: State,
        noWifiAccessPointsState: State
) : DatabaseEntity() {

    val smsId = latState.smsId
    val timestamp = latState.timestamp

    val lat = latState.value
    val lng = lngState.value
    val acc = accState.value

    private val daysSinceLastState = getDaysSinceState(latState)
    private val noCellTowers = noCellTowersState.value.toInt()
    private val noWifiAccessPoints = noWifiAccessPointsState.value.toInt()

    val args: Bundle
        get() = LocationBundle()
                .putDouble(KEY.LAT, lat)
                .putDouble(KEY.LNG, lng)
                .putDouble(KEY.ACC, acc)
                .putDaysSinceLastState(daysSinceLastState)
                .putInt(KEY.NO_CELL_TOWERS, noCellTowers)
                .putInt(KEY.NO_WIFI_ACCESS_POINTS, noWifiAccessPoints)
                .get()

    internal class LocationBundle {

        private val bundle = Bundle()

        fun get(): Bundle = bundle

        fun putDouble(key: KEY, value: Double): LocationBundle = apply {
            bundle.putDouble(key.get(), value)
        }

        fun putDaysSinceLastState(value: Double): LocationBundle = apply {
            bundle.putDouble(DateUtils.DAYS_SINCE_LAST_STATE, value)
        }

        fun putInt(key: KEY, value: Int): LocationBundle = apply {
            bundle.putInt(key.get(), value)
        }
    }

    override val nullType: DatabaseEntity?
        get() = null

    override fun createReportTitle(): String? = null

    override fun createReport(): String? = null

}